File: Validate Binary Search Tree/Validate Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    private boolean recur(TreeNode root, int min, int max){
        if(root == null){
            return true;
        }
        else{
            if((root.val >= max) || (root.val <= min)){
                return false;
            }
            else{
                return recur(root.left, min, root.val) && recur(root.right, root.val, max);
            }
        }
    }
    
    public boolean isValidBST(TreeNode root) {
        return recur(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
}

File: Validate Binary Search Tree/Validate Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    # @param root, a tree node
    # @return a boolean
    def helper(self, root, minVal, maxVal):
        if root is None:
            return True
        elif root.val >= maxVal or root.val <= minVal:
            return False
        else:
            return self.helper(root.left, minVal, root.val) and self.helper(root.right, root.val, maxVal)
    
    def isValidBST(self, root):
        return self.helper(root, -2147483647, 2147483647)
